### PR TITLE
Feature/tao 5364 taskqueue ui component

### PR DESCRIPTION
**Pull request by ssipasseuth**
Requires https://github.com/oat-sa/tao-core/pull/1558

Added the new UI components for the new task queue visualisation.

For task queue setup doc, check https://github.com/oat-sa/extension-tao-task-queue readme or single use the following commands:
1. Activate async task queue: `sudo -u www-data php index.php 'oat\taoTaskQueue\scripts\tools\InitializeQueue' --broker=rds --persistence=default --receive=10`
2. To start and stop the worker that will listen and process the task queue : `sudo -u www-data php index.php 'oat\taoTaskQueue\scripts\tools\RunWorker'`

### CODE REVIEW

File: controller/TaskQueueWebApi.php [hunk 1]
```diff
@@ -22,8 +22,13 @@
...
 use oat\oatbox\filesystem\FileSystemService;
-use oat\taoTaskQueue\model\Entity\CategoryEntityDecorator;
-use oat\taoTaskQueue\model\TaskLog\CategoryCollectionDecorator;
+use oat\taoTaskQueue\model\Entity\Decorator\CategoryEntityDecorator;
+use oat\taoTaskQueue\model\Entity\Decorator\HasFileEntityDecorator;
+use oat\taoTaskQueue\model\QueueDispatcherInterface;
+use oat\taoTaskQueue\model\TaskLog\Decorator\CategoryCollectionDecorator;
+use oat\taoTaskQueue\model\TaskLog\Decorator\FieldRemoverCollectionDecorator;
+use oat\taoTaskQueue\model\TaskLog\Decorator\HasFileCollectionDecorator;
+use oat\taoTaskQueue\model\TaskLogBroker\TaskLogBrokerInterface;
```
Comment (ionutpad) [inline controller/TaskQueueWebApi.php]: Not used import

File: model/TaskLog/Decorator/HasFileCollectionDecorator.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    public function __construct(TaskLogCollectionInterface $collection, FileSystemService $fileSystemService)
+    {
+        parent::__construct($collection);
+
+        $this->fileSystemService = $fileSystemService;
+    }
+
+    /**
+     * CAUTION: Parent not used, so this decorator needs to be the first one in the queue.
```
Comment (ionutpad) [inline model/TaskLog/Decorator/HasFileCollectionDecorator.php]: Nobody it's going to read this :)
I think we need another approach to not go afterwards reading this comment.

File: views/js/component/listing/list.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+ * Copyright (c) 2017 (original work) Open Assessment Technologies SA ;
+ */
+define([
+    'jquery',
+    'lodash',
+    'i18n',
+    'ui/component',
+    'taoTaskQueue/component/listing/element',
+    'tpl!taoTaskQueue/component/listing/tpl/list',
+    'tpl!taoTaskQueue/component/listing/tpl/elementWrapper',
```
Comment (no-chris) [inline views/js/component/listing/list.js:25]: What is the purpose of the element wrapper? Why the element cannot render itself with its own id?

File: views/js/component/manager/manager.js [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+], function ($, _, __, hider, component, badgeFactory, makeAlignable, makeAbsorbable, makePulsable, listElementFactory, reportElementFactory, taskListFactory, managerTpl) {
+    'use strict';
+
+    /**
+     * Transform the task log summary into a configuration set for the badge
+     * @param {Object} tasksStatuses - the task log summary
+     * @returns {Object} the new badge data to be displayed following the format {type, loading, value}
+     */
+    var getBadgeDataFromStatus = function getBadgeDataFromStatus(tasksStatuses){
+        var running = (tasksStatuses.numberOfTasksInProgress > 0);
```
Comment (no-chris) [inline views/js/component/manager/manager.js]: this holds a boolean thus should be prefixed with `is`: `isRunning`

File: views/js/component/manager/manager.js [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        if(tasksStatuses){
+            if(tasksStatuses.numberOfTasksFailed){
+                return {
+                    type : 'error',
+                    loading : running,
```
Comment (no-chris) [inline views/js/component/manager/manager.js]: same here: `isLoading`

File: views/js/model/taskQueueModel.js [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+                                    model.trigger('singletaskadded', taskData);
+                                }else if(!hasSameState(_cache[taskData.id], taskData)){
+                                    //check if the status has changed
+                                    model.trigger('singletaskstatuschange', taskData);
+                                }
+                            }else{
+                                _cache = {};
+                            }
+                            _cache[taskData.id] = taskData;
+                            return Promise.resolve(taskData);
```
Comment (no-chris) [inline views/js/model/taskQueueModel.js]: you can simply `return taskData`, as the return value of `.then()` is automatically converted into a resolved Promise

File: views/js/model/taskQueueModel.js [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+                                });
+                            }else{
+                                _.forEach(taskData, function(task){
+                                    newCache[task.id] = task;
+                                });
+                            }
+                            //update local cache
+                            _cache = newCache;
+
+                            return Promise.resolve(taskData);
```
Comment (no-chris) [inline views/js/model/taskQueueModel.js]: same here

File: controller/TaskQueueWebApi.php [hunk 8]
```diff
@@ -73,15 +77,26 @@ public function getAll()
...
         }
 
-        if ($this->hasRequestParameter(self::PARAMETER_REPORT_INCLUDED)) {
-            $reportIncluded = (bool) $this->getRequestParameter(self::PARAMETER_REPORT_INCLUDED);
-        }
+        /** @var FileSystemService $fs */
+        $fs = $this->getServiceManager()->get(FileSystemService::SERVICE_ID);
+
+        // adding hasFile to the results
+        $collection = new HasFileCollectionDecorator(
```
Comment (ionutpad) [inline controller/TaskQueueWebApi.php]: I think this collection approach can have performance issues. (for each collection you iterate and only add one value, as more collection iterators we have as slow this can became).
If you remember we discuss last PR to have a general CollectionDecorator who receives entities decorators and just apply the decorate for each entity.
You can achieve  this easy:
Have a CollectionDecorator which will be constructed with $data, and array (collection) of entity decorators. when method toArray() called the array data will be transformed to entities decorated with the decorators.
This means that the search method will return TaskLogCollectionInterface which it's extended by CollectionDecorator.
So basically the search will return as well a collection in your case will be CollectionDecorator. 
If you pass no decorators to the CollectionDecorator the outcome will be the same (default behavior).

File: views/js/model/taskQueueModel.js [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+            archive : function archive(taskId){
+
+                var status;
+
+                if(!config.url || !config.url.archive){
+                    throw new TypeError('config.url.archive is not configured while archive() is being called');
+                }
+
+                status = request(config.url.archive, {taskId : taskId}, 'GET', {}, true)
+                    .then(function(){
```
Comment (no-chris) [inline views/js/model/taskQueueModel.js]: This callback is not necessary.

File: views/js/model/taskQueueModel.js [hunk 10]
```diff
@@ -0,0 +1,481 @@
...
+                                    result.extra = creationResult.extra;
+                                }
+                                if(result.finished){
+                                    //send to queue
+                                    self.trigger('fastFinished', result);
+                                }else{
+                                    //send to queue
+                                    self.trigger('enqueued', result);
+                                }
+                                return Promise.resolve(result);
```
Comment (no-chris) [inline views/js/model/taskQueueModel.js]: `return result`

File: views/js/test/component/manager/test.js [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+            id: 'rdf#i1508337970199318643',
+            taskName: 'Task Name',
+            taskLabel: 'Task label',
+            status: 'completed',
+            owner: 'userId',
+            createdAt: '1512120107',
+            updatedAt: '1512121107',
+            createdAtElapsed : 601,
+            updatedAtElapsed :26,
+            hasFile: false,//suppose
```
Comment (no-chris) [inline views/js/test/component/manager/test.js]: we have weird comments...

File: views/js/test/component/manager/test.js [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+            taskName: 'Task Name 2',
+            taskLabel: 'Task label 2',
+            status: 'in_progress',
+            owner: 'userId',
+            createdAt: '1512122107',
+            updatedAt: '1512123107',
+            createdAtElapsed : 41,
+            updatedAtElapsed :626,
+            hasFile: false,
+            category: 'publish',//d
```
Comment (no-chris) [inline views/js/test/component/manager/test.js]: ... here...

File: views/js/test/component/manager/test.js [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+            createdAt: '1512124107',
+            updatedAt: '1512125107',
+            createdAtElapsed : 61,
+            updatedAtElapsed :101,
+            hasFile: true,//suppose
+            category: 'export',//d
+            report : {
+                type : 'error',
+                message : 'running task rdf#i1508337970190342',
+                data : null,//download url ? task context ?
```
Comment (no-chris) [inline views/js/test/component/manager/test.js]: ... and up to there!

File: views/js/component/listing/list.js (hunk 3 above)
Comment (ssipasseuth) [reply views/js/component/listing/list.js:25]: This is to potentially reuse the element outside of a list. The element wrapper is but to wrap the element inside a `li` element.

**Review by ionutpad — APPROVED** [review]

File: views/js/component/button/taskable.js [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2017 (original work) Open Assessment Technologies SA;
+ */
+
+/**
+ * Allow to generate an absorbing animation from a target element to the component
```
Comment (no-chris) [inline views/js/component/button/taskable.js]: Wrong comment and examples? ;)

File: views/js/component/button/taskable.js [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+
+                if (result.finished) {
+                    if(task.hasFile){
+                        //download if its is a export-typed task
+                        taskQueue.download(task.id).then(function(){
+                            //immediately archive the finished task as there is no need to display this task in the queue list
+                            taskQueue.archive(task.id).then(function () {
+                                self.trigger('finished', result);
+                                taskQueue.pollAll();
+                            });
```
Comment (no-chris) [inline views/js/component/button/taskable.js]: You should chain Promises here:
```
taskQueue.download(task.id).then(function(){
    return taskQueue.archive(task.id);
}).then(function () {
    self.trigger('finished', result);
    taskQueue.pollAll();
});
```

File: views/js/component/button/treeButton.js [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+            if(this.is('started')){
+                this.setState('started', false);
+                this.trigger('stop');
+            }
+            return this;
+        }
+    };
+
+    /**
+     * Create a button with to create a task
```
Comment (no-chris) [inline views/js/component/button/treeButton.js]: with... what ?! ;-)

File: views/js/component/manager/manager.js [hunk 17]
```diff
@@ -293,6 +298,16 @@ define([
                     e.stopPropagation();
                 });
 
+                //close the popup when clicking outside of the component
+                $(document).off('click.task-queue-manager').on('click.task-queue-manager', function(e){
+                    if($trigger.get(0) !== e.target && !$.contains($trigger.get(0), e.target)){
```
Comment (no-chris) [inline views/js/component/manager/manager.js]: When is this component rendered? If on page load it is a bit heavy to have this listener always on in `document`. Maybe you could set the listener in the `.show()` method and remove it in the `.hide()` method ?

Comment (ssipasseuth) [reply views/js/component/manager/manager.js]: Good point !
